File: backend/app/services/clinical_support.py
"""Clinical decision support: suggested tests, specialist, urgency, follow-up."""
from __future__ import annotations

from typing import Any

DISCLAIMER = (
    "This decision support is for educational and research purposes only "
    "and does not replace professional medical advice."
)

_SUPPORT: dict[str, dict[str, Any]] = {
    "diabetes": {
        "specialist": "Endocrinologist",
        "suggested_tests": ["HbA1c", "Fasting Blood Sugar", "Oral Glucose Tolerance Test", "Lipid Profile"],
        "follow_up": "Recheck glucose and HbA1c in 3 months.",
    },
    "heart": {
        "specialist": "Cardiologist",
        "suggested_tests": ["ECG", "Echocardiogram", "Lipid Profile", "Blood Pressure Monitoring"],
        "follow_up": "Cardiology consultation and stress evaluation if symptoms persist.",
    },
    "kidney": {
        "specialist": "Nephrologist",
        "suggested_tests": ["Serum Creatinine", "eGFR", "Urine Albumin", "Blood Urea"],
        "follow_up": "Repeat renal panel in 4–6 weeks; monitor blood pressure.",
    },
    "liver": {
        "specialist": "Hepatologist / Gastroenterologist",
        "suggested_tests": ["Liver Function Test", "Bilirubin Test", "SGPT / SGOT", "Abdominal Ultrasound"],
        "follow_up": "Repeat LFT in 4 weeks; viral hepatitis screening as clinically indicated.",
    },
}


def urgency_from_risk(risk_level: str) -> str:
    r = (risk_level or "").lower()
    if r in {"high"}:
        return "Within 7 days"
    if r in {"moderate", "medium"}:
        return "Within 30 days"
    return "Routine (6–12 months)"


def clinical_support(disease: str, risk_level: str) -> dict[str, Any]:
    base = _SUPPORT.get(disease.lower(), {
        "specialist": "General Physician",
        "suggested_tests": ["General health screening"],
        "follow_up": "Routine review.",
    })
    return {
        "disease": disease,
        "specialist": base["specialist"],
        "suggested_tests": base["suggested_tests"],
        "follow_up": base["follow_up"],
        "urgency": urgency_from_risk(risk_level),
        "priority_reason": f"{risk_level} risk for {disease}",
        "disclaimer": DISCLAIMER,
    }

File: src/lib/clinical-support.ts
/**
 * Clinical Decision Support — deterministic rules per disease.
 * Mirrors backend `/clinical-support` shape; works fully offline.
 */
import type { DiseaseKey, PredictionResult } from "./predict-api";

export type Urgency = "Routine (6–12 months)" | "Within 30 days" | "Within 7 days" | "Immediate";

export type ClinicalSupport = {
  disease: DiseaseKey;
  specialist: string;
  suggested_tests: string[];
  follow_up: string;
  urgency: Urgency;
  priority_reason: string;
  disclaimer: string;
};

export type AppointmentRecommendation = {
  specialist: string;
  priority: Urgency;
  reason: string;
};

export const CLINICAL_DISCLAIMER =
  "This system provides AI-based risk estimates for educational and research purposes only. It is not a medical diagnosis and should not replace professional consultation.";

const SPECIALIST: Record<DiseaseKey, string> = {
  diabetes: "Endocrinologist",
  heart: "Cardiologist",
  kidney: "Nephrologist",
  liver: "Hepatologist / Gastroenterologist",
};

const TESTS: Record<DiseaseKey, string[]> = {
  diabetes: ["HbA1c", "Fasting Blood Sugar", "Oral Glucose Tolerance Test", "Lipid Profile"],
  heart: ["ECG", "Echocardiogram", "Lipid Profile", "Blood Pressure Monitoring"],
  kidney: ["Serum Creatinine", "eGFR", "Urine Albumin", "Blood Urea"],
  liver: ["Liver Function Test", "Bilirubin Test", "SGPT / SGOT", "Abdominal Ultrasound"],
};

const FOLLOW_UP: Record<DiseaseKey, string> = {
  diabetes: "Recheck glucose and HbA1c in 3 months.",
  heart: "Cardiology consultation and stress evaluation if symptoms persist.",
  kidney: "Repeat renal panel in 4–6 weeks; monitor blood pressure.",
  liver: "Repeat LFT in 4 weeks; viral hepatitis screening as clinically indicated.",
};

export function urgencyFromRisk(risk: PredictionResult["risk"] | string): Urgency {
  const r = String(risk).toLowerCase();
  if (r === "high") return "Within 7 days";
  if (r === "medium" || r === "moderate") return "Within 30 days";
  return "Routine (6–12 months)";
}

export function getClinicalSupport(disease: DiseaseKey, risk: PredictionResult["risk"]): ClinicalSupport {
  return {
    disease,
    specialist: SPECIALIST[disease],
    suggested_tests: TESTS[disease],
    follow_up: FOLLOW_UP[disease],
    urgency: urgencyFromRisk(risk),
    priority_reason: `${risk.charAt(0).toUpperCase() + risk.slice(1)} risk for ${disease}`,
    disclaimer: CLINICAL_DISCLAIMER,
  };
}

export function getAppointmentRecommendation(
  disease: DiseaseKey,
  risk: PredictionResult["risk"],
): AppointmentRecommendation {
  return {
    specialist: SPECIALIST[disease],
    priority: urgencyFromRisk(risk),
    reason: `${risk.charAt(0).toUpperCase() + risk.slice(1)} ${disease} risk detected by AI screening.`,
  };
}
